test/module/shared_model/backend_proto/shared_proto_tx_response_test.cpp

```diff
@@ -46,3 +48,32 @@ TEST(ProtoTxResponse, TxResponseLoad) {
               shared_model::crypto::Hash(hash));
   });
 }
+
+/**
+ * @given TransactionResponse that previously had lazy fields
+ * @when those lazy fields are simultaneously accessed
+ * @then there is no race condition and segfaults
+ */
+TEST(TxResponse, SafeToReadFromMultipleThreads) {
+  iroha::protocol::ToriiResponse response;
+  const std::string hash = "123";
+  response.set_tx_hash(hash);
+  response.set_tx_status(iroha::protocol::TxStatus::COMMITTED);
+  auto model_response = shared_model::proto::TransactionResponse(response);
+
+  const auto num_threads = 20;
+
+  auto multiple_access = [&model_response] {
+    // old good way to cause race condition on lazy fields
+    ASSERT_TRUE(model_response == model_response);
```

Comment:
Is it legal to use `ASSERT_EQ` here?

Reply:
It looks like we can substitute ASSERT_TRUE with ASSERT_EQ, but ASSERT_TRUE was intentionally used here to be independent of the implementation of ASSERT_EQ. Usage of  ASSERT_TRUE can make us sure that operator== is called.

Reply:
What else can be called here?

Reply:
Evaluation order can be mangled by gtest stuff. We need to cause fields initialization as much simultaneously as it is possible